crates/wasi-http/src/body.rs

```diff
@@ -0,0 +1,513 @@
+                    // explicitly drop the body_writer so that anything waiting on the read end returns
+                    // immediately.
+                    drop(body_writer);
+
+                    let trailers = frame.into_trailers().unwrap();
+
+                    // TODO: this will fail in two cases:
+                    // 1. we've already used the channel once, which should be imposible,
+                    // 2. the read end is closed.
+                    // I'm not sure how to differentiate between these two cases, or really
+                    // if we need to do anything to handle either.
+                    let _ = trailer_writer.send(Ok(trailers));
+
+                    break;
+                }
+
+                assert!(frame.is_data(), "frame wasn't data");
+
+                let data = frame.into_data().unwrap();
+
+                // If the receiver no longer exists, thats ok - in that case we want to keep the
+                // loop running to relieve backpressure, so we get to the trailers.
+                let _ = body_writer.send(Ok(data)).await;
+            }
+        });
+
+        Self {
+            worker,
+            stream: Some(HostIncomingBodyStream::new(body_receiver)),
+            trailers,
+        }
+    }
+}
+
+pub struct HostFutureTrailers {
+    pub worker: AbortOnDropJoinHandle<()>,
+    pub received: Option<Result<FieldMap, types::Error>>,
+    pub receiver: oneshot::Receiver<Result<hyper::HeaderMap, anyhow::Error>>,
+}
+
+impl HostFutureTrailers {
+    pub fn ready(&mut self) -> impl std::future::Future<Output = anyhow::Result<()>> + '_ {
+        use std::future::Future;
+        use std::task::{Context, Poll};
+
+        struct TrailersReady<'a>(&'a mut HostFutureTrailers);
+
+        impl<'a> Future for TrailersReady<'a> {
+            type Output = anyhow::Result<()>;
+
+            fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                if self.0.received.is_some() {
+                    return Poll::Ready(Ok(()));
+                }
+
+                match Pin::new(&mut self.0.receiver).poll(cx) {
+                    Poll::Ready(Ok(Ok(headers))) => {
+                        self.0.received = Some(Ok(FieldMap::from(headers)))
+                    }
+
+                    Poll::Ready(Ok(Err(e))) => {
+                        self.0.received = Some(Err(types::Error::ProtocolError(format!(
+                            "hyper error: {e:?}"
+                        ))))
+                    }
+
+                    Poll::Ready(Err(_)) => {
+                        self.0.received = Some(Err(types::Error::ProtocolError(
+                            "stream hung up before trailers were received".to_string(),
+                        )))
+                    }
+
+                    Poll::Pending => return Poll::Pending,
+                }
+
+                Poll::Ready(Ok(()))
+            }
+        }
+
+        TrailersReady(self)
+    }
```

Comment:
This looks like it could be `async fn ready` from my reading, but was there a reason that wasn't chosen?

Reply:
Because this is a oneshot receiver, the `receiver` gets moved as part of the await in an `async fn`. The impl future avoids that without having to jump through any weird hoops for cancel safety.

Reply:
explained in 47232f3ce

Reply:
Ah makes sense! I've attempted a refactor using `impl Future for &mut T` in https://github.com/bytecodealliance/wasmtime/pull/7073 which I think may work to help clean this up a bit

Reply:
thanks, I never thought of that but of course that is helpful.